runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -713,6 +705,86 @@ private void emitResults() {
     }
   }
 
+  private DoFnRunner<InputT, OutputT> ensureStateCleanup(
+      SdkHarnessDoFnRunner<InputT, OutputT> sdkHarnessRunner) {
+    if (keyCoder == null) {
+      // There won't be any state to clean up
+      // (stateful functions have to be keyed)
+      return sdkHarnessRunner;
+    }
+    // Takes care of state cleanup via StatefulDoFnRunner
+    Coder windowCoder = windowingStrategy.getWindowFn().windowCoder();
+    StatefulDoFnRunner.CleanupTimer<InputT> cleanupTimer =
+        new StatefulDoFnRunner.CleanupTimer<InputT>() {
+
+          private static final String GC_TIMER_ID = "__user-state-cleanup__";
+
+          @Override
+          public Instant currentInputWatermarkTime() {
+            return timerInternals.currentInputWatermarkTime();
+          }
+
+          @Override
+          public void setForWindow(InputT input, BoundedWindow window) {
+            Preconditions.checkNotNull(input, "Null input passed to CleanupTimer");
+            // make sure this fires after any window.maxTimestamp() timers
+            Instant gcTime = LateDataUtils.garbageCollectionTime(window, windowingStrategy).plus(1);
+            ByteBuffer key;
+            try {
+              key =
+                  ByteBuffer.wrap(
+                      CoderUtils.encodeToByteArray((Coder) keyCoder, ((KV) input).getKey()));
+            } catch (CoderException e) {
+              throw new RuntimeException("Failed to encode key for Flink state backend", e);
+            }
+            // Ensure the state backend is not concurrently accessed by the state requests
+            try {
+              stateBackendLock.lock();
+              // Set these two to ensure correct timer registration
+              // 1) For the timer setting
+              sdkHarnessRunner.setCurrentTimerKey(key);
+              // 2) For the timer deduplication
+              getKeyedStateBackend().setCurrentKey(key);
+              timerInternals.setTimer(
+                  StateNamespaces.window(windowCoder, window),
+                  GC_TIMER_ID,
+                  gcTime,
```

Comment:
Should we compare gcTime in case of pre-existing timer for a key?

Reply:
The state namespace is assumed to be unique for each window, if not then they are the same window and the GC time should be the same. Internally, `StateNamespaces$WindowNamespace#stringKey()` serializes the Window to a byte array and base64 encodes it to a string. That's a bit crazy but that's how Beam state works for windows.